File: frontend/apps/ios/Pantopus/Features/Shared/Wizard/Blocks/UploadSlotsBlock.swift
//
//  UploadSlotsBlock.swift
//  Pantopus
//
//  Wizard content block — vertical stack of upload tiles. Used by P20's
//  claim ownership Step 2 to collect ID + proof-of-ownership documents.
//

import SwiftUI

/// Render-state for one tile.
public enum UploadSlotState: Sendable, Equatable {
    case empty
    case picked(name: String, sizeBytes: Int)
    case uploading(name: String, fraction: Double)
    case uploaded(name: String, sizeBytes: Int)
    case failed(name: String, message: String)
}

/// One upload-tile descriptor passed to `UploadSlotsBlock`.
public struct UploadSlot: Sendable, Equatable, Identifiable {
    public let id: String
    public let title: String
    public let acceptHint: String
    public let state: UploadSlotState

    public init(id: String, title: String, acceptHint: String, state: UploadSlotState) {
        self.id = id
        self.title = title
        self.acceptHint = acceptHint
        self.state = state
    }
}

/// 2-up vertical stack of upload tiles. Each tile is 160 pt tall with
/// a dashed border + sunken background; tap fires `onPick` and the
/// caller drives state transitions through the consumer's view-model.
@MainActor
public struct UploadSlotsBlock: View {
    private let slots: [UploadSlot]
    private let onPick: @MainActor (String) -> Void
    private let onRemove: @MainActor (String) -> Void

    public init(
        slots: [UploadSlot],
        onPick: @escaping @MainActor (String) -> Void,
        onRemove: @escaping @MainActor (String) -> Void = { _ in }
    ) {
        self.slots = slots
        self.onPick = onPick
        self.onRemove = onRemove
    }

    public var body: some View {
        VStack(spacing: Spacing.s3) {
            ForEach(slots) { slot in
                UploadTile(slot: slot, onPick: onPick, onRemove: onRemove)
            }
        }
    }
}

private struct UploadTile: View {
    let slot: UploadSlot
    let onPick: @MainActor (String) -> Void
    let onRemove: @MainActor (String) -> Void

    var body: some View {
        Button(action: tapHandler) {
            content
                .frame(maxWidth: .infinity)
                .frame(height: 160)
                .background(Theme.Color.appSurfaceSunken)
                .clipShape(RoundedRectangle(cornerRadius: Radii.xl, style: .continuous))
                .overlay(border)
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(a11y)
        .accessibilityIdentifier("uploadSlot_\(slot.id)")
        .overlay(alignment: .topTrailing) {
            if showsRemove {
                Button {
                    onRemove(slot.id)
                } label: {
                    Icon(.x, size: 16, color: Theme.Color.appTextInverse)
                        .frame(width: 28, height: 28)
                        .background(Theme.Color.appText.opacity(0.7))
                        .clipShape(Circle())
                        .padding(Spacing.s2)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Remove \(slot.title)")
                .accessibilityIdentifier("uploadSlot_remove_\(slot.id)")
            }
        }
    }

    private func tapHandler() {
        switch slot.state {
        case .empty, .failed:
            onPick(slot.id)
        default:
            break
        }
    }

    private var showsRemove: Bool {
        switch slot.state {
        case .picked, .uploaded, .failed: true
        case .empty, .uploading: false
        }
    }

    @ViewBuilder private var content: some View {
        switch slot.state {
        case .empty:
            emptyState
        case let .picked(name, size), let .uploaded(name, size):
            filledState(name: name, size: size, isUploaded: isUploaded)
        case let .uploading(name, fraction):
            uploadingState(name: name, fraction: fraction)
        case let .failed(name, message):
            failedState(name: name, message: message)
        }
    }

    private var emptyState: some View {
        VStack(spacing: Spacing.s2) {
            Icon(.upload, size: 28, color: Theme.Color.appTextSecondary)
            Text(slot.title)
                .pantopusTextStyle(.body)
                .foregroundStyle(Theme.Color.appText)
            Text("Tap to upload")
                .pantopusTextStyle(.caption)
                .foregroundStyle(Theme.Color.appTextSecondary)
            Text(slot.acceptHint)
                .font(.system(size: 11))
                .foregroundStyle(Theme.Color.appTextMuted)
        }
        .padding(Spacing.s4)
    }

    private func filledState(name: String, size: Int, isUploaded: Bool) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            HStack(spacing: Spacing.s2) {
                Icon(isUploaded ? .checkCircle : .file,
                     size: 22,
                     color: isUploaded ? Theme.Color.success : Theme.Color.primary600)
                VStack(alignment: .leading, spacing: 2) {
                    Text(slot.title)
                        .pantopusTextStyle(.caption)
                        .foregroundStyle(Theme.Color.appTextSecondary)
                    Text(name)
                        .pantopusTextStyle(.body)
                        .foregroundStyle(Theme.Color.appText)
                        .lineLimit(2)
                }
                Spacer()
            }
            Text(formatSize(size))
                .pantopusTextStyle(.caption)
                .foregroundStyle(Theme.Color.appTextSecondary)
        }
        .padding(Spacing.s4)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func uploadingState(name: String, fraction: Double) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s3) {
            HStack(spacing: Spacing.s2) {
                ProgressView()
                Text("Uploading \(name)…")
                    .pantopusTextStyle(.body)
                    .foregroundStyle(Theme.Color.appText)
                    .lineLimit(1)
            }
            ProgressView(value: max(0.05, min(1.0, fraction)))
                .tint(Theme.Color.primary600)
        }
        .padding(Spacing.s4)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func failedState(name: String, message: String) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            HStack(spacing: Spacing.s2) {
                Icon(.alertCircle, size: 22, color: Theme.Color.error)
                Text(name)
                    .pantopusTextStyle(.body)
                    .foregroundStyle(Theme.Color.appText)
            }
            Text(message)
                .pantopusTextStyle(.caption)
                .foregroundStyle(Theme.Color.error)
            Text("Tap to retry")
                .pantopusTextStyle(.caption)
                .foregroundStyle(Theme.Color.primary600)
        }
        .padding(Spacing.s4)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var border: some View {
        RoundedRectangle(cornerRadius: Radii.xl, style: .continuous)
            .strokeBorder(
                style: StrokeStyle(
                    lineWidth: borderWidth,
                    lineCap: .round,
                    dash: borderDash
                )
            )
            .foregroundStyle(borderColor)
    }

    private var borderColor: Color {
        switch slot.state {
        case .empty: Theme.Color.appBorderStrong
        case .picked, .uploading: Theme.Color.primary600
        case .uploaded: Theme.Color.success
        case .failed: Theme.Color.error
        }
    }

    private var borderWidth: CGFloat {
        switch slot.state {
        case .empty, .failed: 1.5
        default: 1
        }
    }

    private var borderDash: [CGFloat] {
        switch slot.state {
        case .empty, .failed: [6, 4]
        default: []
        }
    }

    private var isUploaded: Bool {
        if case .uploaded = slot.state { return true }
        return false
    }

    private var a11y: String {
        switch slot.state {
        case .empty:
            "\(slot.title). Tap to upload. \(slot.acceptHint)"
        case let .picked(name, _):
            "\(slot.title). \(name) selected. Tap remove to reset."
        case let .uploading(name, fraction):
            "\(slot.title). Uploading \(name), \(Int(fraction * 100)) percent."
        case let .uploaded(name, _):
            "\(slot.title). \(name) uploaded."
        case let .failed(name, message):
            "\(slot.title). Failed to upload \(name). \(message). Tap to retry."
        }
    }

    private func formatSize(_ bytes: Int) -> String {
        let mb = Double(bytes) / 1_048_576.0
        if mb >= 1 { return String(format: "%.1f MB", mb) }
        let kb = Double(bytes) / 1_024.0
        return String(format: "%.0f KB", kb)
    }
}

#Preview {
    UploadSlotsBlock(
        slots: [
            UploadSlot(id: "id", title: "Government ID", acceptHint: "JPG, PNG, or PDF up to 10 MB", state: .empty),
            UploadSlot(id: "doc", title: "Proof of ownership", acceptHint: "JPG, PNG, or PDF up to 10 MB", state: .uploaded(name: "Deed.pdf", sizeBytes: 256_000))
        ],
        onPick: { _ in },
        onRemove: { _ in }
    )
    .padding()
    .background(Theme.Color.appBg)
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/wizard/blocks/UploadSlotsBlock.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.shared.wizard.blocks

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/** Render-state for one tile. */
sealed interface UploadSlotState {
    data object Empty : UploadSlotState

    data class Picked(val name: String, val sizeBytes: Long) : UploadSlotState

    data class Uploading(val name: String, val fraction: Float) : UploadSlotState

    data class Uploaded(val name: String, val sizeBytes: Long) : UploadSlotState

    data class Failed(val name: String, val message: String) : UploadSlotState
}

/** One upload-tile descriptor passed to [UploadSlotsBlock]. */
data class UploadSlot(
    val id: String,
    val title: String,
    val acceptHint: String,
    val state: UploadSlotState,
)

/**
 * Vertical stack of upload tiles. Each tile is 160 dp tall with a dashed
 * border (empty/failed) or solid border (picked/uploading/uploaded) and a
 * sunken background; tap fires [onPick] when empty/failed and the X-overlay
 * fires [onRemove] when there's a file.
 */
@Composable
fun UploadSlotsBlock(
    slots: List<UploadSlot>,
    onPick: (String) -> Unit,
    modifier: Modifier = Modifier,
    onRemove: (String) -> Unit = {},
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        slots.forEach { slot ->
            UploadTile(slot = slot, onPick = onPick, onRemove = onRemove)
        }
    }
}

@Suppress("LongMethod")
@Composable
private fun UploadTile(
    slot: UploadSlot,
    onPick: (String) -> Unit,
    onRemove: (String) -> Unit,
) {
    val borderColor = slot.state.borderColor()
    val isClickable = slot.state is UploadSlotState.Empty || slot.state is UploadSlotState.Failed
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .height(160.dp)
                .clip(RoundedCornerShape(Radii.xl))
                .background(PantopusColors.appSurfaceSunken)
                .border(
                    width = if (slot.state is UploadSlotState.Empty || slot.state is UploadSlotState.Failed) 1.5.dp else 1.dp,
                    color = borderColor,
                    shape = RoundedCornerShape(Radii.xl),
                )
                .let { mod -> if (isClickable) mod.clickable { onPick(slot.id) } else mod }
                .testTag("uploadSlot_${slot.id}")
                .semantics { contentDescription = slot.state.a11y(slot) },
    ) {
        when (val state = slot.state) {
            is UploadSlotState.Empty -> EmptyContent(slot)
            is UploadSlotState.Picked ->
                FilledContent(slot = slot, name = state.name, sizeBytes = state.sizeBytes, isUploaded = false)
            is UploadSlotState.Uploaded ->
                FilledContent(slot = slot, name = state.name, sizeBytes = state.sizeBytes, isUploaded = true)
            is UploadSlotState.Uploading -> UploadingContent(name = state.name, fraction = state.fraction)
            is UploadSlotState.Failed -> FailedContent(name = state.name, message = state.message)
        }
        if (slot.state.showsRemove) {
            Box(
                modifier =
                    Modifier
                        .align(Alignment.TopEnd)
                        .padding(Spacing.s2)
                        .size(28.dp)
                        .clip(CircleShape)
                        .background(PantopusColors.appText.copy(alpha = 0.7f))
                        .clickable { onRemove(slot.id) }
                        .testTag("uploadSlot_remove_${slot.id}"),
                contentAlignment = Alignment.Center,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.X,
                    contentDescription = "Remove ${slot.title}",
                    size = 16.dp,
                    tint = PantopusColors.appTextInverse,
                )
            }
        }
    }
}

@Composable
private fun EmptyContent(slot: UploadSlot) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(Spacing.s4),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s2, Alignment.CenterVertically),
    ) {
        PantopusIconImage(
            icon = PantopusIcon.Upload,
            contentDescription = null,
            size = 28.dp,
            tint = PantopusColors.appTextSecondary,
        )
        Text(slot.title, style = PantopusTextStyle.body, color = PantopusColors.appText)
        Text("Tap to upload", style = PantopusTextStyle.caption, color = PantopusColors.appTextSecondary)
        Text(slot.acceptHint, style = PantopusTextStyle.caption, color = PantopusColors.appTextMuted)
    }
}

@Composable
private fun FilledContent(
    slot: UploadSlot,
    name: String,
    sizeBytes: Long,
    isUploaded: Boolean,
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(Spacing.s4),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            PantopusIconImage(
                icon = if (isUploaded) PantopusIcon.CheckCircle else PantopusIcon.File,
                contentDescription = null,
                size = 22.dp,
                tint = if (isUploaded) PantopusColors.success else PantopusColors.primary600,
            )
            Column(modifier = Modifier.fillMaxWidth()) {
                Text(slot.title, style = PantopusTextStyle.caption, color = PantopusColors.appTextSecondary)
                Text(
                    name,
                    style = PantopusTextStyle.body,
                    color = PantopusColors.appText,
                    fontWeight = FontWeight.SemiBold,
                )
            }
        }
        Text(formatSize(sizeBytes), style = PantopusTextStyle.caption, color = PantopusColors.appTextSecondary)
    }
}

@Composable
private fun UploadingContent(
    name: String,
    fraction: Float,
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(Spacing.s4),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            CircularProgressIndicator(modifier = Modifier.size(18.dp), strokeWidth = 2.dp)
            Text("Uploading $name…", style = PantopusTextStyle.body, color = PantopusColors.appText)
        }
        LinearProgressIndicator(
            progress = { fraction.coerceIn(0.05f, 1f) },
            modifier = Modifier.fillMaxWidth(),
            color = PantopusColors.primary600,
        )
    }
}

@Composable
private fun FailedContent(
    name: String,
    message: String,
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(Spacing.s4),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            PantopusIconImage(
                icon = PantopusIcon.AlertCircle,
                contentDescription = null,
                size = 22.dp,
                tint = PantopusColors.error,
            )
            Text(name, style = PantopusTextStyle.body, color = PantopusColors.appText)
        }
        Text(message, style = PantopusTextStyle.caption, color = PantopusColors.error)
        Spacer(Modifier.height(2.dp))
        Text("Tap to retry", style = PantopusTextStyle.caption, color = PantopusColors.primary600)
    }
}

private fun UploadSlotState.borderColor(): Color =
    when (this) {
        is UploadSlotState.Empty -> PantopusColors.appBorderStrong
        is UploadSlotState.Picked, is UploadSlotState.Uploading -> PantopusColors.primary600
        is UploadSlotState.Uploaded -> PantopusColors.success
        is UploadSlotState.Failed -> PantopusColors.error
    }

private val UploadSlotState.showsRemove: Boolean
    get() =
        when (this) {
            is UploadSlotState.Picked, is UploadSlotState.Uploaded, is UploadSlotState.Failed -> true
            is UploadSlotState.Empty, is UploadSlotState.Uploading -> false
        }

private fun UploadSlotState.a11y(slot: UploadSlot): String =
    when (this) {
        is UploadSlotState.Empty -> "${slot.title}. Tap to upload. ${slot.acceptHint}"
        is UploadSlotState.Picked -> "${slot.title}. $name selected. Tap remove to reset."
        is UploadSlotState.Uploading -> "${slot.title}. Uploading $name, ${(fraction * 100).toInt()} percent."
        is UploadSlotState.Uploaded -> "${slot.title}. $name uploaded."
        is UploadSlotState.Failed -> "${slot.title}. Failed to upload $name. $message. Tap to retry."
    }

private fun formatSize(bytes: Long): String {
    val mb = bytes.toDouble() / 1_048_576.0
    if (mb >= 1) return "%.1f MB".format(mb)
    val kb = bytes.toDouble() / 1_024.0
    return "%.0f KB".format(kb)
}
